### ISSUE TITLE
Implement a crawler to build a migration database

### ISSUE THREAD

**Issue by tiagostutz** [bancodobrasil/stop-analyzing-api#5]
I was wondering if we should implement a mock data generator or build a database from a real data and I think the latter is a better approach because real data guides better the design decisions, here and at the UI.

So, it would be nice to crawl data from the customer website that inspired this project and will be the first one to use: https://www.lafiancee.com.br/vestidos-de-noiva

I know that crawling in Golang is not the best approach, so you can build the crawler on your own user on GitHub using your prefered stack and we will work on the issues there. 

The constraints:
- it should run on docker container and expose a volume that will store the data as a JSON
- it should expose a HTTP endpoint to download this data as a zip file

The ideia is that this service will be composed with this service here to be used as a migration to the database

What do you think?

**Comment by douglasferlini** [comment]
Great, I can work on it.

**Comment by tiagostutz** [comment]
Awesome! 
Assigned to you my friend.

As soon as you create the repo on your user, please paste the link here so we can help with the discussion and issues there.

**Comment by tiagostutz** [comment]
Hey, @douglasferlini  just for you to know, I just updated some commiting guidelines here https://github.com/bancodobrasil/stop-analyzing-api/blob/master/CONTRIBUTING.md#commiting-and-pr-guidelines

I think it would be good for you to notice that and if you have any suggestion, please feel free to open an issue with that.

**Comment by douglasferlini** [comment]
I've already created the fork https://github.com/douglasferlini/stop-analyzing-api on my user.

**Comment by douglasferlini** [comment]
Which data we will be extracting from this website? Just the dresses and their attributes?

**Comment by tiagostutz** [comment]
Yes, the dresses, their attributes and theirs pictures. The attributes are the ones that you can filter using this filter:

<img width="232" alt="Captura de Tela 2020-07-08 às 07 18 16" src="https://user-images.githubusercontent.com/3986989/86907345-407cf500-c0eb-11ea-81d0-0a4894a8d0df.png">


So, turns out that the engineering design of API has evolved a bit - as you can see on #8 - and this crawler won't be part of stop-analyzing-api (this repo). You should create a new repo on your user for this project and you can build it in any language you prefer (including GO, of course). Also, I suggest you to make this project a [Poppins](https://github.com/bancodobrasil/poppins) project, following [this guidelines](https://github.com/bancodobrasil/poppins/blob/master/MANIFESTO.md), so it will be a good project for first time contributors. We can even ask for some contributors here to take a sneak there. I can help you with this, also.

As soon as you create the crawler repo, comment here with link to this repo and we will help you to manage that. Sounds good?

**Comment by douglasferlini** [comment]
Great, I will create a repo stop-analyzing-crawler. Is this name ok?

**Comment by tiagostutz** [comment]
Hi @douglasferlini ! Sorry for the late response... I had an atypical week, but now I'm breathing again... eheheheheh...

Maybe the name could be driven to the data source name (lafiancee-website). I think you could name it **lafiancee-website-crawler**. This data you are going to crawl can be used anywhere, not just with Stop Analyzing, that's why I think you could use another naming approach. Got it?

**Comment by douglasferlini** [comment]
Hi,

I've created the repository https://github.com/douglasferlini/lafiancee-website-crawler for this project.
As soon as I can I will follow the guidelines to convert it to a Poppins project.

**Comment by douglasferlini** [comment]
Hi @tiagostutz ,

I've already implemented an initial version of the crawler.

**Comment by tiagostutz** [comment]
> Hi @tiagostutz ,
> 
> I've already implemented an initial version of the crawler.

Great!

Do you want some help on that? If so, we can ask some people here whether they would like to contribute.

**Comment by vinnicius** [comment]
> Do you want some help on that? If so, we can ask some people here whether they would like to contribute.

I can help! Gonna fav the crawler here and we can work in issues there

**Comment by douglasferlini** [comment]
> > Hi @tiagostutz ,
> > I've already implemented an initial version of the crawler.
> 
> Great!
> 
> Do you want some help on that? If so, we can ask some people here whether they would like to contribute.

I would like to discuss briefly if the produced json has all that attributes needed.

**Comment by tiagostutz** [comment]
@douglasferlini can you post here the produced JSON? Or are you asking how this format will be?

**Comment by douglasferlini** [comment]
@tiagostutz Currently the json for each dress is being produced as shown below:

{
	"id": "77c441ce-202d-4a9e-1a8f-6ee56d7b16f4",
	"name": "La Fianc\u00e9e Queenie",
	"price": 0,
	"media": [{
		"fullUrl": "https://static.wixstatic.com/media/daf591_1140ec8164e24f44a793272d95dc3c66~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
	}, {
		"fullUrl": "https://static.wixstatic.com/media/daf591_8178d1e7a58144b4a6015d5d13ba1b49~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
	}, {
		"fullUrl": "https://static.wixstatic.com/media/daf591_c1f1937e6e20428ca26c145cb36ac4d4~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
	}],
	"options": [{
		"option": [{
			"value": "Decote em V"
		}, {
			"value": "Manga comprida"
		}]
	}, {
		"option": [{
			"value": "Boho Chic"
		}]
	}, {
		"option": [{
			"value": "Corte Reto"
		}]
	}, {
		"option": [{
			"value": "Bordado"
		}, {
			"value": "Renda"
		}]
	}]
}


Is this contract correct?

**Comment by tiagostutz** [comment]
Hi @douglasferlini..! 

I think this exercise shown that maybe we have a problem with our Model... and that's because we are just considerating "Tags" but it seems we will have to have "Groups of tags". See, the `options` fields brings this information in groups, e.g.: "Decote em V" and "Manga comprida" will be tags in the "Corpete" group. "Boho Chic" is a tag in the "Estilo" group. "Corte Reto" will be a tag in the "modelagem" group:

<img width="551" alt="Captura de Tela 2020-07-24 às 08 59 46" src="https://user-images.githubusercontent.com/3986989/88389038-10408200-cd8c-11ea-94c3-14c3e4a3225f.png">


But maybe if we put a prefix in the tag like: `modelagem-corte-reto`, `estilo-boho-chic`, that would work also. The thing is: will tag with prefix have the same narrow effect than grouping tags in terms of converting the user choice?

I'd like to hear some thoughts from some friends if possible: @flaviostutz @GreenStage @zamariola

**Comment by vinnicius** [comment]
Is there a problem with something like this? 

```json
{
    "id": "77c441ce-202d-4a9e-1a8f-6ee56d7b16f4",
    "name": "La Fianc\u00e9e Queenie",
    "price": 0,
    "media": [{
        "fullUrl": "https://static.wixstatic.com/media/daf591_1140ec8164e24f44a793272d95dc3c66~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
    }, {
        "fullUrl": "https://static.wixstatic.com/media/daf591_8178d1e7a58144b4a6015d5d13ba1b49~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
    }, {
        "fullUrl": "https://static.wixstatic.com/media/daf591_c1f1937e6e20428ca26c145cb36ac4d4~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
    }],
    "options": [{
        "Corpete": ["Decote em V", "Manga comprida"]
    }, {
        "Estilo": ["Boho Chic"]
    }, {
        "Modelagem": ["Corte Reto"]
    }, {
        "Tecido": ["Bordado", "Renda"]
    }]
}
```

**Comment by eniovalo** [comment]
Sorry to ask, but there is a problem if we group all options in the same group?
Example:
```json
"options": [
	"Decote em V",
	"Manga comprida",
	"Boho Chic",
	"Corte Reto",
	"Bordado",
	"Renda"
]
```
Because if tag (option) is unique we can still filter it. =]

**Comment by zamariola** [comment]
Interesting topics... I think that we may need to change our current tag model or keep it as-is (for another purpose may be) and insert a new concept, something like search facets. 

I'm thinking in two options:

1) 
In a traditional relational FK-PK, I'd suggest something as:

Product/Choice <-> Facet (option in our scenarios above)   ===> This relationship shows that our product has its attribute
Facet <-> FacetValue1 ===> This relationship shows that that attribute may have that value
Facet <-> FacetValue2 ===> This relationship shows that that attribute may have that value
Facet <-> FacetValue 3 ===> This relationship shows that that attribute may have that value
Product/Choice <-> FacetValue ===> This relationship shows that that our product has this value select for that attribute

In that case we'd end up with:

La Fiancee Queenie <-> Corpete
Corpete <-> Decote em V
Corpete <-> Manga comprida
La Fiancee Queenie <-> Decote em V

Is that too much or too normalized or too soon?

2)
However, I think we can indeed stick with the prefix and I'd suggest a very uncommon char as a separator, something like one pipe |  or two pipes ||.

Thanks!!

**Comment by GreenStage** [comment]
@zamariola  
I like your approach, I would just make a small change

As @eniovalo said, if the tag is unique in the whole db scope, they don't need to be under a group. It seems to me that having tags listed under "tag-groups" is just a front-end problem.

If so, I don't think the product needs to reference the tag-group directly.

I would suggest something like this

![Untitled](https://user-images.githubusercontent.com/11062001/88454917-4d7b4180-ce6a-11ea-8786-9144dccd3e16.png)

Which implies a server response like:
```
{
   "id":"77c441ce-202d-4a9e-1a8f-6ee56d7b16f4",
   "name":"La Fianc\u00e9e Queenie",
   "price":0,
   "options":[
      {
         "id":"250a2fa4-7b67-449c-a5f1-f5b80fd6d62f",
         "name":"Decote em V",
         "group-id":"25ebdaf2-87f1-41fa-bb95-0a20cfdd06b4"
      },
      {
         "id":"5c6d0c44-11eb-4d93-9eab-87c6c01f0c86",
         "name":"Manga comprida",
         "group-id":"25ebdaf2-87f1-41fa-bb95-0a20cfdd06b4"
      },
      {
         "id":"3f59a726-531c-455b-8e0e-f6682c51fcce",
         "name":"Boho Chic",
         "group-id":"e522407e-c4af-463e-84a5-3f91bdf71a02"
      }
   ],
   "options-groups":[
      {
         "id":"25ebdaf2-87f1-41fa-bb95-0a20cfdd06b4",
         "name":"Corpete"
      },
      {
         "id":"e522407e-c4af-463e-84a5-3f91bdf71a02",
         "name":"Estilo"
      }
   ]
}
```

then let the browser do the grouping of options/tags

**Comment by tiagostutz** [comment]
Really good points.!

As I can see, we reach the decision point of whether we are going with a Tag approach or with an Attribute (we called it Tag-Group) approach in the model. Correct?

I was thinking a lot about how the data will be used to narrow the choices options until we reach the best ones to the users and I think I finally have a simple enough explanation to it. I think this can help us make a decision here.

My current idea is to use a "filter by example" approach in this narrowing algorithm. So, what we are going to do is to take the sequence of choices, get its Tags or Attributes and use them as filter options to evaluate the next choice. Of course we will have to add some randomness in the process, but thats the basic ideia. If the users chooses a dress with "Decote em V" in 3 sequences, even though the other options are all diferent, the "Decote em V" is going to be a filter criteria. The user will make his/her choices until the system has enough filter criteria reinforcement to reach "an end". We will discuss this whole thing in an appropriate issue, but that's the basic concepts and I think they can help us make a decision.

That being said, I vote for the @GreenStage result model but I think we could change the name "Tag" to "AttributeValue" and the "TagGroup" to "Attribute". I think this would be good because, this way, we leave the "Tag" keyword free to be used in a future, in a more appropriated context, like @zamariola pointed out.

What do you think?
As soon as we make the decision, we can discuss the crawler returnin model, as @vinnicius suggested above.

-------

/please-review-my-english

_PS - this /please-review-my-english "command" is something new we are testing which consists in asking volunteers to review your message and give suggestions on how to improve it. The idea is to make open source a great opportunity to learn english as well. The volunteer will review only the "tagged" message and not the entire thread._

**If you want you english to be reviewed, just place this command (`/please-review-my-english`) in your message**

**Comment by MarcioMad** [comment]
Really good points.!

We now reached the opportunity point to define the best way to go:  The Tags approach, or  Attribute (we called it Tag-Group) approach in the model. Do you agree?

I was thinking a lot about how the data will be used to narrow the choices options until we reach the best ones to the users and I think I finally have a simple enough explanation to it. I think this can help us make a decision here.

My current idea is to use a "filter by example" approach in this narrowing algorithm. So, what we are going to do is to take the sequence of choices, get its Tags or Attributes and use them as filter options to evaluate the next choice. Of course we will have to add some randomness to the process, but that is the basic idea. If the user chooses a dress with "low-cut cleavage" in 3 sequences, even though the other options are all different, the "low-cut cleavage" is going to be a filter criteria. The user will make his/her choices until the system has enough filter criteria reinforcement to reach "an end". We will discuss this whole thing in an appropriate issue, but that would be the basic concept and I think it can help us in making a decision.

That said, I vote for the @GreenStage result model but I think we could change the name "Tag" to "AttributeValue" and the "TagGroup" to "Attribute". IMO, this would be good, because, we would leave the "Tag" keyword free to be used in a future, in a more appropriated context, like @zamariola pointed out.

What do you think?
As soon as we make the decision, we can discuss the crawler returnin model, as @vinnicius suggested above.

**Comment by MarcioMad** [comment]
Hi @tiagostutz, 

Where would be the right place to post suggested corrections for "Please-review-my-english" marked text? Is it ok to just comment below the original post?

**Comment by MarcioMad** [comment]
Hi @tiagostutz,

About the proposed method: Would it be like a cascading filter selection? If so, both approaches (tags and attributes) would apply. Even free (tokenized) text would work. I don't think I grasped the whole concept here.

**Comment by tiagostutz** [comment]
> Hi @tiagostutz,
> 
> Where would be the right place to post suggested corrections for "Please-review-my-english" marked text? Is it ok to just comment below the original post?

Yes, it's OK to comment below the original post.! Thanks!

**Comment by tiagostutz** [comment]
> Hi @tiagostutz,
> 
> About the proposed method: Would it be like a cascading filter selection? If so, both approaches (tags and attributes) would apply. Even free (tokenized) text would work. I don't think I grasped the whole concept here.

Yes, it will be a cascading filter selection.  And yes, you are correct, both approaches will work. And, no, you are no being  grasped at all my friend!

My vote for the attribute based approach is that it will give us more information about the data and can expand our UX possibilities. If the UX engineer somehow wants to show to the user which attribute he/she has already decided or which attribute he wants to skip in the choice process, etc, it would be possible. And, with this attribute based approach, if needed, we could derivate the tag process, correct?

@MarcioMad what about opening an issue to start this cascade filter algorithm? Let's go for it?

**Comment by tiagostutz** [comment]
@GreenStage what do you think about the suggestion of changing from TagGroup/Tag to Attribute/AttributeValue ?

@andreaguiar-br what are your thoughts on this? Maybe you could share a dbdiagram with some suggestion?

**Comment by GreenStage** [comment]
@tiagostutz cheers, i don't think i understand the problem.

From what I could gather its just a naming issue, right?

**Comment by tiagostutz** [comment]
Yep, just a naming issue.

**Comment by GreenStage** [comment]
Hmm, it feels to me like AttributeValue / Attribute suggest a 1-1 relationship

**Comment by tiagostutz** [comment]
> Hmm, it feels to me like AttributeValue / Attribute suggest a 1-1 relationship
> 
> 
> 
> Edit: it doesn't, it does suggest a one-to-many, but in a way that one attribute can only have one value at a given time in the front end.
> 
> 
> 
> If this is the case, I agree!

Thats exactly the case! Nice.!

Let’s wait until monday to see if someone has more suggestion/contributions on this topic and we end up with this decision.

Cheers !

**Comment by tiagostutz** [comment]
So, I think we have reached a good point here and the JSON suggested by @vinnicius can be used to extract data and record on the database using the model we just discussed here. In the JSON, the only thing to change is the `options` property name to `attributes`. 

The input/output would be as follows:

**JSON produced by the Crawler**:
```
{
    "id": "77c441ce-202d-4a9e-1a8f-6ee56d7b16f4",
    "name": "La Fiancèe Queenie",
    "price": 0,
    "media": [{
        "fullUrl": "https://static.wixstatic.com/media/daf591_1140ec8164e24f44a793272d95dc3c66~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
    }, {
        "fullUrl": "https://static.wixstatic.com/media/daf591_8178d1e7a58144b4a6015d5d13ba1b49~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
    }, {
        "fullUrl": "https://static.wixstatic.com/media/daf591_c1f1937e6e20428ca26c145cb36ac4d4~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
    }],
    "attributes": [{
        "Corpete": ["Decote em V", "Manga comprida"]
    }, {
        "Estilo": ["Boho Chic"]
    }, {
        "Modelagem": ["Corte Reto"]
    }, {
        "Tecido": ["Bordado", "Renda"]
    }]
}
```

This JSON will be recorded on the database this way:

**Product:**
| id | title | contentURL | active |
|---|-----|-------------|--------|
| 1  | La Fiancèe Queenie | https://static.wixstatic.com/media/daf591_1140ec8164e24f44a793272d95dc3c66~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg | true |

**Attribute:**
| id | text |
|---|-----|
| 1  | Corpete |
| 2  | Estilo |
| 3  | Modelagem |
| 4 | Tecido |

**AttributeValue:**
| id | attribute_id | value |
|---|-------------|------|
| 1  | 1 | Decote em V |
| 2 | 1 | Manga comprida |
| 3 | 2 | Boho Chic |
| 4 | 3 | Corte Reto |
| 5 | 4 | Bordado |
| 6 | 4 | Renda |

**Product_AttributeValue**
| id | product_id | attribute_value_id |
|---|-------------|------|
| 1  | 1 | 1 |
| 2  | 1 | 2 |
| 3  | 1 | 3 |
| 4  | 1 | 4 |
| 5  | 1 | 5 |
| 6  | 1 | 6 |

@douglasferlini do you think you can continue with these definitions?

**Comment by andreaguiar-br** [comment]
> @GreenStage what do you think about the suggestion of changing from TagGroup/Tag to Attribute/AttributeValue ?
> 
> @andreaguiar-br what are your thoughts on this? Maybe you could share a [dbdiagram](https://dbdiagram.io) screenshot/URL with some suggestion?

@tiagostutz , Sorry my late response...

I really liked the model @GreenStage had suggested and also the meaning of new fields, changed from _Tag_ to _Attribute_, fits very good. 

Once data migration is quite expensive, when I plan a database, I think It´s crucial to understand the context of information, and in this case, the information named _Attribute_ really fit a characteristic of a type of product. and could include also size, color, etc. 

A way I usually do to name the information I can't control, like this case, is take a portion of data (as much as I can), analyze different values and try to fit in real world and create some definition, and check in the future to make sure it keeps the definition.

Hope I had helped..

**Comment by douglasferlini** [comment]
> So, I think we have reached a good point here and the JSON suggested by @vinnicius can be used to extract data and record on the database using the model we just discussed here. In the JSON, the only thing to change is the `options` property name to `attributes`. 
> 
> The input/output would be as follows:
> 
> **JSON produced by the Crawler**:
> ```
> {
>     "id": "77c441ce-202d-4a9e-1a8f-6ee56d7b16f4",
>     "name": "La Fiancèe Queenie",
>     "price": 0,
>     "media": [{
>         "fullUrl": "https://static.wixstatic.com/media/daf591_1140ec8164e24f44a793272d95dc3c66~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
>     }, {
>         "fullUrl": "https://static.wixstatic.com/media/daf591_8178d1e7a58144b4a6015d5d13ba1b49~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
>     }, {
>         "fullUrl": "https://static.wixstatic.com/media/daf591_c1f1937e6e20428ca26c145cb36ac4d4~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg"
>     }],
>     "attributes": [{
>         "Corpete": ["Decote em V", "Manga comprida"]
>     }, {
>         "Estilo": ["Boho Chic"]
>     }, {
>         "Modelagem": ["Corte Reto"]
>     }, {
>         "Tecido": ["Bordado", "Renda"]
>     }]
> }
> ```
> 
> This JSON will be recorded on the database this way:
> 
> **Product:**
> | id | title | contentURL | active |
> |---|-----|-------------|--------|
> | 1  | La Fiancèe Queenie | https://static.wixstatic.com/media/daf591_1140ec8164e24f44a793272d95dc3c66~mv2.jpg/v1/fit/w_500,h_500,q_90/file.jpg | true |
> 
> **Attribute:**
> | id | text |
> |---|-----|
> | 1  | Corpete |
> | 2  | Estilo |
> | 3  | Modelagem |
> | 4 | Tecido |
> 
> **AttributeValue:**
> | id | attribute_id | value |
> |---|-------------|------|
> | 1  | 1 | Decote em V |
> | 2 | 1 | Manga comprida |
> | 3 | 2 | Boho Chic |
> | 4 | 3 | Corte Reto |
> | 5 | 4 | Bordado |
> | 6 | 4 | Renda |
> 
> **Product_AttributeValue**
> | id | product_id | attribute_value_id |
> |---|-------------|------|
> | 1  | 1 | 1 |
> | 2  | 1 | 2 |
> | 3  | 1 | 3 |
> | 4  | 1 | 4 |
> | 5  | 1 | 5 |
> | 6  | 1 | 6 |
> 
> @douglasferlini do you think you can continue with these definitions?

Yes, I will change the produced JSON.

**Comment by douglasferlini** [comment]
The produced JSON is already adjusted to match the model mentioned above.

Please, let me know if there is any other issues.

Warm Regards

**Comment by tiagostutz** [comment]
@douglasferlini awesome!!

Have you fully ran the crawler? If so, could you zip the JSON and share it with us using some cloud file sharing service, please?

**Comment by douglasferlini** [comment]
> @douglasferlini awesome!!
> 
> Have you fully ran the crawler? If so, could you zip the JSON and share it with us using some cloud file sharing service, please?

The JSON is available on http://35.198.4.79:8000/data.json .
